File: library/src/main/java/com/pengrad/telegrambot/request/SendAnimation.java
package com.pengrad.telegrambot.request;

import com.pengrad.telegrambot.model.MessageEntity;
import com.pengrad.telegrambot.model.request.ParseMode;

import java.io.File;

/**
 * Stas Parshin
 * 28 July 2018
 */
public class SendAnimation extends AbstractMultipartRequest<SendAnimation> {

    public SendAnimation(Object chatId, String animation) {
        super(chatId, animation);
    }

    public SendAnimation(Object chatId, File animation) {
        super(chatId, animation);
    }

    public SendAnimation(Object chatId, byte[] animation) {
        super(chatId, animation);
    }

    public SendAnimation duration(int duration) {
        return add("duration", duration);
    }

    public SendAnimation width(int width) {
        return add("width", width);
    }

    public SendAnimation height(int height) {
        return add("height", height);
    }

    /**
     * @deprecated Use thumbnail instead
     */
    @Deprecated 
    public SendAnimation thumb(byte[] thumb) {
        return thumbnail(thumb);
    }

    /**
     * @deprecated Use thumbnail instead
     */
    @Deprecated 
    public SendAnimation thumb(File thumb) {
        return thumbnail(thumb);
    }

    public SendAnimation thumbnail(byte[] thumbnail) {
        return super.thumbnail(thumbnail);
    }

    public SendAnimation thumbnail(File thumbnail) {
        return super.thumbnail(thumbnail);
    }

    public SendAnimation caption(String caption) {
        return add("caption", caption);
    }

    public SendAnimation parseMode(ParseMode parseMode) {
        return add("parse_mode", parseMode.name());
    }

    public SendAnimation captionEntities(MessageEntity... entities) {
        return add("caption_entities", entities);
    }

    public SendAnimation hasSpoiler(boolean has_spoiler) {
        return add("has_spoiler", has_spoiler);
    }

    public SendAnimation showCaptionAboveMedia(Boolean showCaptionAboveMedia) {
        return add("show_caption_above_media", showCaptionAboveMedia);
    }

    @Override
    protected String getFileParamName() {
        return "animation";
    }

    @Override
    public String getDefaultFileName() {
        return ContentTypes.GIF_FILE_NAME;
    }

    @Override
    public String getDefaultContentType() {
        return ContentTypes.GIF_MIME_TYPE;
    }
}

File: library/src/main/java/com/pengrad/telegrambot/request/SendAnimation.kt
package com.pengrad.telegrambot.request

import com.pengrad.telegrambot.model.MessageEntity
import com.pengrad.telegrambot.model.request.ParseMode
import com.pengrad.telegrambot.utility.kotlin.checkDeprecatedConstructorParameters
import com.pengrad.telegrambot.utility.kotlin.optionalRequestParameter
import java.io.File

class SendAnimation private constructor(
    chatId: Long? = null,
    channelUsername: String? = null,

    animationUrl: String? = null,
    animationFile: File? = null,
    animationBytes: ByteArray? = null
) : AbstractThumbnailedMultipartRequest<SendAnimation>(
    chatId = chatId,
    channelUsername = channelUsername,

    contentParameterName = "animation",
    contentUrl = animationUrl,
    contentFile = animationFile,
    contentBytes = animationBytes,

    defaultFileName = ContentTypes.GIF_FILE_NAME,
    defaultContentType = ContentTypes.GIF_MIME_TYPE
) {


    constructor(chatId: Long, animationUrl: String) : this(
        chatId = chatId,
        channelUsername = null,
        animationUrl = animationUrl
    )

    constructor(channelUsername: String, animationUrl: String) : this(
        chatId = null,
        channelUsername = channelUsername,
        animationUrl = animationUrl
    )


    constructor(chatId: Long, animationFile: File) : this(
        chatId = chatId,
        channelUsername = null,
        animationFile = animationFile
    )

    constructor(channelUsername: String, animationFile: File) : this(
        chatId = null,
        channelUsername = channelUsername,
        animationFile = animationFile
    )


    constructor(chatId: Long, animationBytes: ByteArray) : this(
        chatId = chatId,
        channelUsername = null,
        animationBytes = animationBytes
    )

    constructor(channelUsername: String, animationBytes: ByteArray) : this(
        chatId = null,
        channelUsername = channelUsername,
        animationBytes = animationBytes
    )


    @Deprecated("Use constructor with chatId or channelUsername instead", ReplaceWith("SendAnimation(chatId, animation)"))
    constructor(chatId: Any, animation: String) : this(
        chatId = (chatId as? Number)?.toLong(),
        channelUsername = chatId as? String,
        animationUrl = animation
    ) {
        checkDeprecatedConstructorParameters()
    }

    @Deprecated("Use constructor with chatId or channelUsername instead", ReplaceWith("SendAnimation(chatId, animation)"))
    constructor(chatId: Any, animation: File) : this(
        chatId = (chatId as? Number)?.toLong(),
        channelUsername = chatId as? String,
        animationFile = animation
    ) {
        checkDeprecatedConstructorParameters()
    }

    @Deprecated("Use constructor with chatId or channelUsername instead", ReplaceWith("SendAnimation(chatId, animation)"))
    constructor(chatId: Any, animation: ByteArray) : this(
        chatId = (chatId as? Number)?.toLong(),
        channelUsername = chatId as? String,
        animationBytes = animation
    ) {
        checkDeprecatedConstructorParameters()
    }

    var caption: String? by optionalRequestParameter()
    var parseMode: ParseMode? by optionalRequestParameter()
    var captionEntities: List<MessageEntity>? by optionalRequestParameter()
    var showCaptionAboveMedia: Boolean? by optionalRequestParameter()
    var hasSpoiler: Boolean? by optionalRequestParameter()

    var duration: Int? by optionalRequestParameter()
    var width: Int? by optionalRequestParameter()
    var height: Int? by optionalRequestParameter()

    fun caption(caption: String) = apply { this.caption = caption }

    fun parseMode(parseMode: ParseMode) = apply { this.parseMode = parseMode }

    fun captionEntities(captionEntities: List<MessageEntity>) = apply { this.captionEntities = captionEntities }

    fun captionEntities(vararg captionEntities: MessageEntity) = apply { this.captionEntities = captionEntities.toList() }

    fun showCaptionAboveMedia(showCaptionAboveMedia: Boolean) = apply { this.showCaptionAboveMedia = showCaptionAboveMedia }

    fun hasSpoiler(hasSpoiler: Boolean) = apply { this.hasSpoiler = hasSpoiler }

    fun duration(duration: Int) = apply { this.duration = duration }

    fun width(width: Int) = apply { this.width = width }

    fun height(height: Int) = apply { this.height = height }

}